File: Medium/1631.py
from typing import List
import heapq

class Solution:
  def minimumEffortPath(self, heights: List[List[int]]) -> int:
    # Dijkstra 算法，计算 (0, 0) 到 (m - 1, n - 1) 的最小体力消耗
    return self.dijkstra(heights)

  def dijkstra(self, matrix):
    # 记录从起点 (0, 0) 到每个节点的最小体力消耗
    m = len(matrix)
    n = len(matrix[0])
    dist_to = [[float("inf")] * n for _ in range(m)]
    dirs = [[0, 1], [1, 0], [0, -1], [-1, 0]]

    # pq 里面存的是 (effort_from_start, row, col)
    # Python heapq 默认按照 tuple 第一项排序，所以会按照 effort_from_start 从小到大排序
    pq = []

    # 从起点 (0, 0) 开始进行 dijkstra 算法
    dist_to[0][0] = 0
    heapq.heappush(pq, (0, 0, 0))

    while pq:
      effort_from_start, cur_row, cur_col = heapq.heappop(pq)

      # 已经存在更优路径，则跳过
      if dist_to[cur_row][cur_col] < effort_from_start:
        continue

      # 判断是否已经到达目标点
      # 如果要提前返回，可以打开这一段
      # if cur_row == m - 1 and cur_col == n - 1:
      #   return dist_to[cur_row][cur_col]

      # For Next One
      for nx, ny in dirs:
        next_row = cur_row + nx
        next_col = cur_col + ny

        # 索引越界
        if next_row >= m or next_row < 0 or next_col >= n or next_col < 0:
          continue

        # 从起点到下一个节点的体力消耗 =
        # 从起点到当前节点的体力消耗 和 当前节点到下一个节点的体力消耗 的较大值
        # Math.max 的原因是因为题目要求的是路径上每一步的体力消耗的最大值
        next_effort_from_start = max(effort_from_start,abs(matrix[next_row][next_col] - matrix[cur_row][cur_col]))

        # 存在更优路径，则更新 dist_to 数组和优先级队列
        if dist_to[next_row][next_col] > next_effort_from_start:
          dist_to[next_row][next_col] = next_effort_from_start
          heapq.heappush(pq, (next_effort_from_start, next_row, next_col))

    return dist_to[m - 1][n - 1] if dist_to[m - 1][n - 1] >= 0 else -1

File: Medium/1631.cs
using System;
using System.Collections.Generic;

public class Solution {
  public int MinimumEffortPath(int[][] heights) {
    // Dijkstra 算法，计算 (0, 0) 到 (m - 1, n - 1) 的最小体力消耗
    return Dijkstra(heights);
  }

  // 记录当前位置和从起点到当前位置的最小体力消耗
  public class State {
    public int row;
    public int col;
    public int effortFromStart;

    public State(int row, int col, int effortFromStart) {
      this.row = row;
      this.col = col;
      this.effortFromStart = effortFromStart;
    }
  }

  public int Dijkstra(int[][] matrix) {
    // 记录从起点 (0, 0) 到每个节点的最小体力消耗
    int m = matrix.Length;
    int n = matrix[0].Length;

    int[][] distTo = new int[m][];
    for (int i = 0; i < m; i++) {
      distTo[i] = new int[n];
      for (int j = 0; j < n; j++) {
        distTo[i][j] = int.MaxValue;
      }
    }

    int[][] dirs = new int[][] {
      new int[] { 0, 1 },
      new int[] { 1, 0 },
      new int[] { 0, -1 },
      new int[] { -1, 0 }
    };

    PriorityQueue<State, int> pq = new PriorityQueue<State, int>();

    // 从起点 (0, 0) 开始进行 dijkstra 算法
    distTo[0][0] = 0;
    pq.Enqueue(new State(0, 0, 0), 0);

    while (pq.Count > 0) {
      State state = pq.Dequeue();
      int curRow = state.row;
      int curCol = state.col;
      int curEffortFromStart = state.effortFromStart;

      // 已经存在更优路径，则跳过
      if (distTo[curRow][curCol] < curEffortFromStart) {
        continue;
      }

      // 判断是否已经到达目标点
      // 如果要提前返回，可以打开这一段
      // if (curRow == m - 1 && curCol == n - 1) {
      //   return distTo[curRow][curCol];
      // }

      // For Next One
      foreach (int[] dir in dirs) {
        int nextRow = curRow + dir[0];
        int nextCol = curCol + dir[1];

        // 索引越界
        if (nextRow >= m || nextRow < 0 || nextCol >= n || nextCol < 0) {
          continue;
        }

        // 从起点到下一个节点的体力消耗 =
        // 从起点到当前节点的体力消耗 和 当前节点到下一个节点的体力消耗 的较大值
        // Math.Max 的原因是因为题目要求的是路径上每一步的体力消耗的最大值
        int nextEffortFromStart = Math.Max(curEffortFromStart, Math.Abs(matrix[nextRow][nextCol] - matrix[curRow][curCol]));

        // 存在更优路径，则更新 distTo 数组和优先级队列
        if (distTo[nextRow][nextCol] > nextEffortFromStart) {
          distTo[nextRow][nextCol] = nextEffortFromStart;
          pq.Enqueue(new State(nextRow, nextCol, nextEffortFromStart), nextEffortFromStart);
        }
      }
    }

    return distTo[m - 1][n - 1] >= 0 ? distTo[m - 1][n - 1] : -1;
  }
}